canals/component/component.py

```diff
@@ -95,6 +94,30 @@ def run(self, **kwargs) -> Dict[str, Any]:
         """
 
 
+class ComponentMeta(type):
+    def __call__(cls, *args, **kwargs):
+        # This runs before __new__ is called
+        run_signature = inspect.signature(cls.run)
+        instance = super().__call__(*args, **kwargs)
+        # If the __init__ called component.set_output_types(), __canals_output__ is already populated
+        if not hasattr(instance, "__canals_output__"):
+            # if the run method was decorated, it has a _output_types_cache field assigned
+            instance.__canals_output__ = getattr(instance.run, "_output_types_cache", {})
+        # If the __init__ called component.set_input_types(), __canals_input__ is already populated
+        if not hasattr(instance, "__canals_input__"):
+            instance.__canals_input__ = {
+                # Create the input sockets
+                param: {
+                    "name": param,
+                    "type": run_signature.parameters[param].annotation,
+                    "is_optional": _is_optional(run_signature.parameters[param].annotation),
+                }
+                for param in list(run_signature.parameters)[1:]  # First is 'self' and it doesn't matter.
+            }
+
+        return instance
```

Comment:
[I might have missed something cruclai here, so bear with me.]

Why a metaclass? I see that later, in `_component` you need to even re-create component classes to force them to use this metaclass (https://github.com/deepset-ai/canals/blob/542666a8c629a8136fbe3825a6466acc7cb012c4/canals/component/component.py#L230C9-L230C89). At the same time, this method is doing nothing with the class: it only manipulates the instance.

I normally consider metaclasses really necessary only when one needs to dynamically alter the class. When the subject is the instance I think `__init__` methods are better suited.

Here, in my understanding this would work well as a wrapper of the component's `__init__`, or even as a sort of `__post_init__` if such a thing exists (I haven't checked yet). It could be assigned to the component class in `_component`, just as you do now for the metaclass. Of course we would need to take care of components that have no `__init__` (should be easy). but I would find that overall safer than messing with a dynamic metaclass. Forcing it on a simple component might work, but in the moment where the component class has some other non-trivial class inheritance behavior (or another metaclass of any kind) I think this mechanism is going to break.

Is there anything that wouldn't work this way?

Reply:
The metaclass is needed to hook into the class creation mechanism so I can intercept the instance before the `__init__` is called, call it a `__pre_init__` if you will. Wrapping the `__init__` was my first attempt but isn't viable because you don't know the signature in advance and they're all different, and passing `self, *args, **kwargs` doesn't always work.

A note about the metaclass, I disagree this is "messing", overriding `__call__` is a well-known pattern to inject logic before the `__new__`. Regarding whether a component already has a metaclass, that will be wrapped in turn by our own, nothing breaks in principle, same goes with complex hierarchies. Then of course there are many ways to break a component but I wouldn't say those are related to the metaclass strategy in particular, many things in canals can go wrong if misused.

All in all I think this approach is the best we can get, but I wonder if this complexity is worth it to have a decorator instead of inheriting from a `Component` base class or mixin. I advocated for the decorator in a moment where it seemed we could implement components with simple functions, but it's now clear this will never happen.

Reply:
> The metaclass is needed to hook into the class creation mechanism so I can intercept the instance before the __init__ is called, call it a __pre_init__ if you will.

Ok just to make sure I really understand the code: the `__init__` will get called at line 101, so in fact most of the logic occurs *after* the init. You grab a handle before it, because that's the only option we have, but the logic needs to occur after init was called and all the `set_input/output_types` calls have run. Is this correct?

> Regarding whether a component already has a metaclass, that will be wrapped in turn by our own, nothing breaks in principle, same goes with complex hierarchies

This is the part I'm not sure about. A class can't have more than one metaclass, in any order or level of the hierarchy afaik. Not easily at least. That's also why they're quite rarely used. Conflicting metaclasses normally raise very confusing errors, and they're very hard to work around. I don't say this is a blocker here (this will be a reeeeally rare case indeed), but it is quite advanced and I'm not sure we see all the implications it will carry.

This is the error I'm thinking about:

```
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass 
of the metaclasses of all its bases
```

> I wonder if this complexity is worth it to have a decorator instead of inheriting from a Component base class or mixin.

I also like more the idea of a base class/mixin. I would be happy even if we inject such baseclass dynamically, so that the users won't see it and the existing components won't break. But needs to be tried out.

Reply:
> A class can't have more than one metaclass, in any order or level of the hierarchy afaik

The way I use the component metaclass, yes you can:
```python
In [1]: class Meta(type):
   ...:     def __init__(cls, name, bases, dct):
   ...:         cls.attr = 42
   ...: 

In [2]: class ComponentMeta(type):
   ...:     pass
   ...: 

In [3]: 

In [3]: class MyComponent(metaclass=Meta):
   ...:     pass
   ...: 

In [4]: newclass = ComponentMeta(MyComponent.__name__, MyComponent.__bases__, dict(MyComponent.__dict__))

In [5]: x = newclass()

In [6]: x.attr
Out[6]: 42
```
Don't get me wrong, but I spent quite some time on this topic and I'm not sure this back and forth is productive anymore, let's just search for an alternative:

> I would be happy even if we inject such baseclass dynamically, so that the users won't see it and the existing components won't break.

Let me see what I can do, I'll be back with an alternative 👍

Reply:
If the alternative is more complicated or takes too much time though, let's just take this solution and move on :+1: I'm just making sure we're not overlooking an option that is easier to reason about.